pkg/opts/container.go

```diff
@@ -56,14 +57,36 @@ func WithVolumes(volumeMounts map[string]string) containerd.NewContainerOpts {
 		defer unix.Unmount(root, 0) // nolint: errcheck
 
 		for host, volume := range volumeMounts {
-			if err := copyOwnership(filepath.Join(root, volume), host); err != nil {
-				return err
+			if err := copyExistingContents(filepath.Join(root, volume), host); err != nil {
+				return errors.Wrap(err, "taking runtime copy of volume")
 			}
 		}
 		return nil
 	}
 }
 
+// copyExistingContents copies from the source to the destination and
+// ensures the ownership is appropriately set.
+func copyExistingContents(source, destination string) error {
+	volList, err := ioutil.ReadDir(source)
+	if err != nil {
+		return err
+	}
+	if len(volList) > 0 {
+		srcList, err := ioutil.ReadDir(destination)
+		if err != nil {
+			return err
+		}
+		if len(srcList) == 0 {
+			// If the source volume is empty, copies files from the root into the volume
+			if err := chrootarchive.NewArchiver(nil).CopyWithTar(source, destination); err != nil {
```

Comment:
Just for my education. Why do we need to chroot archive? I don't quite understand it.

Reply:
Mainly just:

> This pulls in and uses github.com/docker/docker/pkg/chrootarchive for the actual copy up which is some battle hardened code to unpack avoiding things like symlink traversal security issues.

This wouldn't be all that hard to implement directly but it does have some slightly tricky corner cases which have already been worked through in that code base. I think the `chroot` aspect of it is just to mitigate the security impact of any bugs.

It also covers a bunch of stuff I wouldn't have thought of myself, such as preserving xattrs and proper handling of hardlinks etc.

Reply:
@ijc Good to know. Thanks for explaining this!